In [1]:
#hiddencell
import pandas as pd 
from pbl_tools import *

# 스테이지 2. 데이터 분석 및 기초 통계량 확인
## 도입
스테이지 2에서는 데이터 분석의 핵심 단계인 기초 통계량 확인에 대해 배웁니다. 이러한 단계는 데이터를 깊이 있게 이해하고 문제 해결에 도움을 주는 중요한 과정입니다.

데이터 분석의 목표는 데이터의 패턴과 특성을 파악하는 것이며, 이를 통해 더 나은 인사이트를 얻고 예측 작업을 진행합니다.

## 학습 목표
- 데이터프레임의 칼럼 이름을 확인하고, 각 변수의 역할을 이해합니다.
- 데이터의 수치형 변수에 대한 기술 통계량을 계산하고 분석합니다. 이를 통해 데이터의 분포와 중요한 통계적 특성을 이해합니다.
- 범주형 변수의 빈도와 다양성을 확인하여 범주형 데이터의 특성을 파악합니다.
- 전처리가 완료된 데이터프레임을 csv 파일로 저장하는 방법에 대해 배웁니다.

# 1. pandas를 이용해 csv 파일 읽어오기
[문제 1]  
`Pandas` 라이브러리(library)를 가져와보세요.  
그리고 `train.csv`, `test.csv`, `sample_submission.csv` 파일을 각각 train, test, submission 변수로 읽어오세요.  
아래 빈칸을 채워주세요.  

In [79]:
import pandas as pd

train = pd.___('train.csv')
test = pd.___('test.csv')
submission = ___.___(___)

In [4]:
#checkcode
ensure_vals(globals(), 'train', 'test', 'submission')
@check_safety
def check(
    user_train: pd.DataFrame,
    user_test: pd.DataFrame,
    user_submission: pd.DataFrame
):
    c_point1 = hasattr(user_train, 'tail')
    c_point2 = hasattr(user_test, 'tail')
    c_point3 = hasattr(user_submission, 'tail')

    if c_point1 and c_point2 and c_point3:
        return True
    else:
        return False

check(train, test, submission)

True

### Inst.  


### Hint.
`pd.read_csv('파일경로/파일명')` 를 활용해 보세요.

### Solution.
```python
import pandas as pd  

train = pd.read_csv('train.csv')  
test = pd.read_csv('test.csv')  
submission = pd.read_csv('sample_submission.csv')
```

# 2. 데이터프레임(DataFrame)정보 출력하기
`info()`함수를 이용해 데이터프레임의 요약 정보를 확인해 봅시다.

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4440 entries, 0 to 4439
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ID              4440 non-null   object
 1   age             4440 non-null   int64 
 2   workclass       3977 non-null   object
 3   fnlwgt          4440 non-null   int64 
 4   education       4440 non-null   object
 5   education.num   4440 non-null   int64 
 6   marital.status  4440 non-null   object
 7   occupation      3972 non-null   object
 8   relationship    4440 non-null   object
 9   race            4440 non-null   object
 10  sex             4440 non-null   object
 11  capital.gain    4440 non-null   int64 
 12  capital.loss    4440 non-null   int64 
 13  hours.per.week  4440 non-null   int64 
 14  native.country  4293 non-null   object
 15  target          4440 non-null   int64 
dtypes: int64(7), object(9)
memory usage: 399.0+ KB


In [82]:
#checkcode
#empty

### Inst.
`info()` 메소드는 pandas 라이브러리(Library)에서 제공하는 함수 중 하나로, **데이터프레임(DataFrame)의 정보를 제공**합니다.  
데이터프레임의 행(row) 수, 칼럼(column) 수, 각 칼럼의 이름, 데이터 타입, 메모리 사용량 등을 보여주며,   
결측치(null)의 개수도 확인할 수 있습니다.    

이 코드 기법은 데이터 분석 및 전처리 과정에서 매우 유용하게 사용됩니다.   
데이터프레임의 크기와 구성을 파악하여 결측치 처리, 데이터 타입 변경, 이상치 처리 등의 작업을 수행할 수 있습니다.   
또한, 메모리 사용량을 파악하여 데이터프레임의 크기를 최적화하는데도 활용됩니다.  

`info()` 메소드의 출력 결과 중에서, 각 칼럼에 대한 정보는 다음과 같은 의미를 가집니다.  

- **column**: 데이터프레임의 **컬럼(열) 이름**을 보여줍니다.
- **Non-null count**: 각 컬럼에 있는 **비어있지 않은 값의 개수**를 보여줍니다.
- **dtype**: 각 컬럼의 **데이터 타입**을 보여줍니다.   
예를 들어, object는 문자열, int는 정수, float는 소수점이 있는 숫자 등을 의미합니다.

---

#### 결과 해석

타겟 변수를 포함하여 총 16개의 피처(feature)와 4440개의 행으로 이루어져 있습니다.   
또한 데이터프레임의 Non-Null Count를 살펴보면 `workclass` 변수에 결측치(누락된 데이터)가 463개,   
`occupation` 변수에는 결측치가 468, `native.country` 변수에 147개가 존재하는 것을 알 수 있습니다.  
이러한 정보를 바탕으로 데이터프레임의 구성과 각 열의 정보를 파악할 수 있으며, 이는 데이터 전처리나 모델링에 필요한 정보를 제공할 수 있습니다.

### Hint.

empty  

### Solution.

empty

# 3. 데이터프레임(DataFrame)의 칼럼(column) 이름 확인하기
[문제 2]  
`columns` 함수를 이용하여 train 데이터셋의 `열(Columns)의 목록`을 확인해 보세요.

In [102]:
train.___

Index(['ID', 'age', 'workclass', 'fnlwgt', 'education', 'education.num',
       'marital.status', 'occupation', 'relationship', 'race', 'sex',
       'capital.gain', 'capital.loss', 'hours.per.week', 'native.country',
       'target'],
      dtype='object')

In [7]:
#checkcode
ensure_vals(globals())
@check_safety
def check(
    user_answer:list,
    number:int
):
    c_point0 = (len(user_answer) == number)
    if c_point0:
        return True
    else:
        return False
check(_,16)

True

### Inst.

pandas 라이브러리를 사용하여 불러온 데이터셋(train)의 **열(column) 이름들을 반환** 하는 속성(attribute)입니다.   
이를 통해, 데이터셋의 **열 이름들을 파악하고, 원하는 열에 쉽게 접근**할 수 있습니다.

pandas DataFrame 객체에서 제공하는 속성으로, 데이터셋의 열 이름들이 포함된 1차원 배열(Series 객체)을 반환합니다.   
이 배열에서는 각 열의 이름이 문자열로 포함되어 있으며, 각 열의 순서는 데이터프레임의 열 순서와 일치합니다.

예를 들어, train 데이터프레임이 'age', 'gender', 'income', 'education' 등의 열(column)을 가지고 있다면,   
train.columns를 호출하면 Index(['age', 'gender', 'income', 'education'], dtype='object')와 같은 결과를 반환합니다.   

### Hint.

empty  

### Solution.

```python
train.columns
```

# 4. 데이터프레임(DataFrame)의 수치형 변수들에 대한 기술 통계량(descriptive statistics) 출력하기
[문제 3]  
`describe()`함수를 사용해 `수치형` 데이터의 통계값을 확인해 봅시다.

In [122]:
train.___()

age        fnlwgt  education.num  capital.gain  capital.loss  \
count  4440.000000  4.440000e+03    4440.000000   4440.000000   4440.000000   
mean     38.779279  1.923122e+05       9.985360   1223.965090     69.511261   
std      14.011257  1.088622e+05       2.580624   8257.227782    357.876753   
min      17.000000  1.882700e+04       1.000000      0.000000      0.000000   
25%      28.000000  1.198022e+05       9.000000      0.000000      0.000000   
50%      37.000000  1.798670e+05      10.000000      0.000000      0.000000   
75%      48.000000  2.402272e+05      12.000000      0.000000      0.000000   
max      90.000000  1.455435e+06      16.000000  99999.000000   2603.000000   

       hours.per.week       target  
count     4440.000000  4440.000000  
mean        39.837387     0.240315  
std         12.760011     0.427323  
min          1.000000     0.000000  
25%         38.000000     0.000000  
50%         40.000000     0.000000  
75%         45.000000     0.000000  
max         99.000000     1.000000

In [9]:
#checkcode
ensure_vals(globals(), 'train')
@check_safety
def check(
    user_answer:pd.DataFrame,
    train_shape:tuple
):
    c_point1 = user_answer.shape == train_shape
    
    if c_point1:
        return True
    else:
        return False
    
check(_, (8,7))

True

### Inst.

`describe()` 함수는 pandas 데이터프레임(DataFrame)에서 **열(column)의 기술 통계량(descriptive statistics)** 을 반환합니다.   
이를 통해 train이라는 데이터프레임에서 **수치형 데이터로 이루어진 열들의 분포와 중심 경향성, 흩어진 정도** 등을 대략적으로 파악할 수 있습니다.   
이 함수는 **결측값(NaN)을 제외** 하고 통계량을 계산하므로, 데이터프레임에 포함된 열들의 실제 분포를 대략적으로 이해하는 데에 도움을 줍니다.

`train.describe()` 함수의 출력 결과를 분석하여, 데이터프레임의 **열들의 분포나 이상치 여부 등을 파악** 할 수 있습니다.  

---

#### 결과 해석
`describe` 함수를 사용하면 각 피처(feature)의 기본적인 **통계 값들을 확인**할 수 있습니다.

- `count`: 해당 피처(feature)의 데이터 **개수**
- `mean`: 해당 피처(feature)의 데이터 **평균값**
- `std`: 해당 피처(feature)의 데이터 **표준편차**
- `min`: 해당 피처(feature)의 데이터 **최솟값**
- `25%, 50%, 75%`: 해당 피처(feature)의 값들을 크기순으로 정렬했을 때, 각각 25%, 50%, 75% 지점의 백분위수
    - 백분위수(percentile)란, 크기가 있는 값들을 순서대로 나열했을 때 그 값의 **상대적인 위치** 를 말합니다.   
    일반적으로 크기가 작은 순으로 오름차순 정렬을 했을 때 0을 최솟값, 100을 최댓값으로 합니다.   
    즉, 100개의 값을 가지는 데이터가 존재할 때 25 백분위수는 그 자료의 값들 중 25번째로 작은 값을 의미합니다.
- `max`: 해당 피처(feature)의 데이터 **최댓값**

각 피처의 count 값이 모두 4440으로 동일하므로 결측치는 없는 것으로 보입니다.   
`fnlwgt` 피처는 상대적으로 큰 표준 편차를 가지고 있으며, 이는 해당 피처의 값들이 평균에서 상당히 퍼져 있음을 나타냅니다.   
또한 `fnlwgt`와 `capital.gain` 피처는 최댓값이 큰 편이므로, 이상치가 존재할 가능성이 있습니다.   
`hours.per.week` 피처의 최솟값이 1로 나타나는데, 이는 이상치의 존재 가능성을 시사합니다.  
마지막으로 `target` 피처는 대부분의 값이 0으로 나타나고, 1의 비율은 24%입니다.  

### Hint.

empty  

### Solution.

```python
train.describe()
```

# 5. 데이터프레임(DataFrame)의 범주형 변수들에 대한 기술 통계량(descriptive statistics) 출력하기
[문제 4]  
- 예측에는 필요하지 않은 `ID` 피처(feature)를 제거해보세요. `drop` 함수는 데이터프레임에서 특정 열을 삭제할 때 사용됩니다.   
여기서 `axis=1` 매개변수는 `열`을 삭제하도록 지정합니다.   
따라서 'ID' 열은 삭제된 데이터프레임(train_drop)에 포함되지 않게 됩니다.
- `describe()` 함수를 활용하여 `범주형 데이터`의 통계적 정보를 확인해보세요.   
`include` 매개변수에 `object`를 할당하면 범주형 데이터의 통계를 살펴볼 수 있습니다.

In [105]:
train_drop = train.drop(['___'],axis = ___) 
train_drop.___(include=['___'])

workclass education      marital.status    occupation relationship  \
count       3977      4440                4440          3972         4440   
unique         8        16                   7            13            6   
top      Private   HS-grad  Married-civ-spouse  Craft-repair      Husband   
freq        2962      1452                2063           543         1790   

         race   sex native.country  
count    4440  4440           4293  
unique      5     2             40  
top     White  Male  United-States  
freq     3761  2975           3928

In [11]:
#checkcode
ensure_vals(globals(), 'train_drop')
@check_safety
def check(
    user_answer:pd.DataFrame,
    train_shape:tuple
):
    c_point1 = user_answer.shape == train_shape
    
    if c_point1:
        return True
    else:
        return False
    
check(_, (4,8))

True

### Inst.
---

#### 결과 해석

결과로 나타난 부분은 범주형 열에 대한 통계 정보입니다.

- `count`: 각 열의 비어 있지 않은(non-null) 데이터의 개수입니다. `workclass` 열에는 463개가 비어있네요.
- `unique`: 각 열에서 고유한(unique) 값의 개수를 나타냅니다. `sex` 열에는 고유한(unique) 값이 2개 있습니다.   
즉, mal과 female 두 가지 값으로 구성되어 있습니다.
- `top`: 각 열에서 가장 자주 나타나는 값(가장 빈번한 값)를 나타냅니다. `native.country` 열에서는 United_States 값의 빈도가 가장 높네요.
- `freq`: 각 열에서 가장 자주 나타나는 값의 빈도수를 나타냅니다. `United_States` 값이 총 3928번 등장했음을 나타냅니다.

### Hint.

empty  

### Solution.

```python
train_drop = train.drop(['ID'],axis = 1) 
train_drop.describe(include=['object'])
```

# 6. submission 데이터의 하위 10개의 행(row) 데이터 출력하기
[문제 5]  
하위 10개의 행(row) 데이터를 출력해 보세요.  

In [134]:
submission.___(___)

ID  target
484  TEST_0484   False
485  TEST_0485   False
486  TEST_0486   False
487  TEST_0487   False
488  TEST_0488   False
489  TEST_0489   False
490  TEST_0490   False
491  TEST_0491   False
492  TEST_0492   False
493  TEST_0493   False

In [13]:
#checkcode
ensure_vals(globals(), 'submission')
@check_safety
def check(
    user_answer:pd.DataFrame,
    submission_shape:tuple
):
    c_point1 = user_answer.shape == submission_shape
    
    if c_point1:
        return True
    else:
        return False
    
check(_, (10,2))

True

### Inst.
`tail()` 함수는 일반적으로 데이터의 **하위 5개 행(row)** 데이터를 출력합니다.   
출력하는 행(row)의 개수를 늘리거나 줄이려면 `tail(n)` 함수의 인자로 원하는 개수를 입력하면 됩니다.

---

#### 결과 해석

제출할 submission 데이터는 **ID와 target(소득)** 칼럼(column)으로 이루어져 있습니다.   
그러나 모든 target 데이터가 0으로 이루어져 있기 때문에,     
이번 프로젝트에서는 **각 ID에 따른 소득을 예측**하는 문제임을 알 수 있습니다.   
**따라서, 예측한 소득으로 target 칼럼(column)의 값을 모두 채우면 됩니다.**  
이를 위해 모델링을 통해 각 ID에 대한 소득을 예측하고,   
이를 target 칼럼에 대입하여 최종적인 submission 데이터를 생성합니다.  

##### **[sample_submission.csv]** 
- ID: 샘플별 고유 ID
- target: 소득 (목표 예측값)
    - 0 : <=50K (5만 달러 이하)  
    - 1 : >50K (5만 달러 초과) 

### Hint.
empty

### Solution.
```python
submission.tail(10)
```

# 7. target의 최빈값(mode) 추출

데이터프레임(train)에서 `target` 빈도수가 가장 높은 값을 출력해보겠습니다.

[문제 6]  
`mode()` 메서드를 이용하여 특정한 열(column)의 `최빈값`을 가져와서 출력해 보세요.

In [151]:
mode_target = train['target'].___()[0]
mode_target

0

In [32]:
#checkcode
ensure_vals(globals(),'mode_target')
@check_safety
def check(
    user_answer:int,
    num:int
):
    c_point0 = user_answer == num
    
    if c_point0:
        return True
    else:
        return False
    
check(mode_target,0)

True

### Inst.
train 데이터의 target 피처(feature)에 대해 `mode()` 함수를 적용하여 `target 피처의 최빈값`을 mode_target 변수에 할당하였습니다.

---
#### 결과 해석
step 4에서 `describe()` 함수의 결과로, target 열에서 값 1이 전체의 24%를 차지한다는 것을 확인했습니다.   
따라서 값 0은 약 76%로 대부분을 차지하게 됩니다.   
이로 인해 최빈값(mode)은 0으로 나타나게 됩니다.   
즉, target 열에서 가장 빈번하게 나타나는 값은 0입니다.   
이것은 데이터셋의 target 값이 주로 0으로 구성되어 있음을 나타냅니다.

### Hint.

empty  

### Solution.

```python
mode_target = train['target'].mode()[0]
mode_target
```

# 8. target의 최빈값으로 submission 데이터 채우기

target 열의 최빈값인 0을 사용하여 submission 파일을 작성한 후, 이를 제출해보겠습니다.

[문제 7]  
submission 데이터의 target 피처(feature)에 `mode_target` 값을 할당해 보세요.

In [165]:
submission['___'] = ___
submission.head()

ID  target
0  TEST_0000       0
1  TEST_0001       0
2  TEST_0002       0
3  TEST_0003       0
4  TEST_0004       0

In [34]:
#checkcode
ensure_vals(globals(),'submission')
@check_safety
def check(
    user_answer:int,
    user_col:int,
    num:int
):
    c_point0 = (user_answer[user_col]).sum() == num
    if c_point0:
        return True
    else:
        return False
check(_,'target',0)

True

### Inst.
이번 프로젝트는 소득을 예측한 값으로 submission 데이터를 채우는 것이 목표라고 했었는데요.  
이번 스텝에서는 소득을 예측한 값으로 채우지 않고, **train 데이터셋에서 구한 target의 최빈값으로 채워보려고 합니다.**  

train 데이터셋에서 구한 target의 최빈값으로 채우는 이유는,   
이전 스텝에서 target 값이 모두 0으로 되어 있는 것을 확인했기 때문입니다.
  
target 열을 합리적으로 예측할 수 있는 값으로 채워서 제출해야 하는데요.  
일반적으로, target 열의 최빈값(mode)을 구해서 이를 채우는 방법이 많이 사용됩니다.     
이렇게 하면 최소한의 예측 성능은 보장됩니다.   
다만, 이를 통해 높은 예측 성능을 얻기 위해서는 전처리나 모델 선택 등의 추가적인 작업이 필요합니다.   
 
해당 작업들은 뒤에서 배울 예정입니다. 
차근차근 배워봅시다!  

이를 위해서는 submission 데이터의 target 피처(feature)에 mode_target 값을 할당하면 됩니다.   
이때, submission['target'] = mode_target와 같이 값을 할당할 수 있습니다.
>- **submission['특성 이름'] = 대체할 값**

### Hint.

empty  

### Solution.
```python
submission['target'] = mode_target
submission.head()
```

# 9.데이터프레임(DataFrame)을 CSV 파일로 저장하기
[문제 8]  
값이 할당된 submission 데이터를 submission.csv 파일로 저장해 보세요.

In [37]:
submission.___('submission.csv', index=False)

In [39]:
#checkcode
ensure_vals(globals(), 'submission')
@check_safety
def check(
    user_answer
):
    answer = pd.read_csv('submission.csv')

    sub_medv = user_answer['target'].apply(lambda x: int(x * 100))
    ans_medv = answer['target'].apply(lambda x: int(x * 100))

    c_point0 = (sub_medv == ans_medv).all()
    if c_point0:
        return True
    else:
        return False

check(submission)

True

### Inst.

### Hint.
`to_csv()` 함수를 사용해서 csv 파일로 저장해 보세요.   
index 인자는 반드시 False로 설정해 주세요.

### Solution.
```python
submission.to_csv('submission.csv', index = False)
```